File: php-poo/aulas/aula13/Cachorro.php
<?php
require_once 'Lobo.php';
class Cachorro extends Lobo
{
    //@Sobrepor
    public function emitirSom()
    {
        echo "<p> Au! Au! Au! </p>";
    }
    function reagirFrase($frase){
        if( ($frase == "Toma Comida") || ($frase == "Olá") ) {
            echo "<p> Abanar e Latir! </p>";
        } else {
            echo "<p> Rosnar! </p>";
        }
    }
    function reagirHoraMinuto($hora, $minuto){
        if($hora < 12) {
            echo "<p> Abanar! </p>";
        } elseif($hora >= 18) {
            echo "<p> Ignorar! </p>";
        } else {
            echo "<p> Abanar e Latir! </p>";
        }
    }
    function reagirDono($dono){
        if($dono == true) {
            echo "<p> Abanar! </p>";
        } else {
            echo "<p> Rosnar e Latir! </p>";
        }
    }
    function reagirIdadePeso($idade, $peso){
        if ($idade < 5) {

            if($peso < 10) {
                echo "<p> Abanar! </p>";
            } else {
                echo "<p> Latir! </p>";
            }

        } else {

            if($peso < 10) {
                echo "<p> Rosnar! </p>";
            } else {
                echo "<p> Ignorar! </p>";
            }
        }
    }

        /* 2. Polimorfismo de Sobrecarga
         * Assinatura Diferentes no Método (Quantidade e os tipos dos parâmetros diferentes);
         * Mesma Classe.
         *
         * Vimos na aula o exemplo da Classe Cachorro - utilizando o metodo reagir()!
         * Ao tentar executar o Polimorfismo Sobrecarga, o PHP apresenta "fatal error:  Cannot redeclare Cachorro::reagir()";
         * O PHP não suporta o Polimorfismo Sobrecarga!
         * Solução simples: "Utilizar funções com nomes diferentes (nomes parecidos) " (dica do prof. Gustavo Guanabara)
         * Criamos:
         * function reagirFrase()
         * function reagirHoraMinuto()
         * function reagirDono()
         * function reagirFrasePeso()
         */
}

File: java-poo/projects/poo-aula13/src/aula13/Cachorro.java
package aula13;

public class Cachorro extends Lobo {

    // Polimorfismo de Sobreposição -> Mesma assinatura. Classes diferentes.
    @Override
    public void emitirSom() {
        System.out.println("Au! Au! Au!");
    }

    // Polimorfismo de Sobrecarga -> mesma classe. Assinatura diferentes.
    public void reagir(String frase) {
        if (frase.toLowerCase() == "toma comida" || frase.toLowerCase() == "ola") {
           System.out.println("Abanar! e Latir!");
        } else {
            System.out.println("Rosnar!");
        }
    }
    public void reagir(int hora, int minuto) {
        if (hora < 12) {
            System.out.println("Abanar!");
        }
        else if (hora >= 18) {
            System.out.println("Ignorar!");
        } else {
            System.out.println("Abanar! e Latir!");
        }
    }
    public void reagir(boolean dono) {
        if (dono) {
            System.out.println("Abanar!");
        } else {
            System.out.println("Rosnar! e Latir!");
        }
    }
    public void reagir(int idade, float peso) {
        if (idade < 5) {
            if (peso < 10){
                System.out.println("Abanar!");
            }
            else {
                System.out.println("Latir!");
            }
        } else {
            if (peso < 10) {
                System.out.println("Rosnar!");
            } else {
                System.out.println("Ignorar!");
            }
        }
    }
}
